pkg/apimachinery/discoverysummarizer/discoverysummarizer.go

```diff
@@ -0,0 +1,166 @@
+	if err != nil {
+		return fmt.Errorf("Failed to bring up the server: %v", err)
+	}
+	return nil
+}
+
+// Handler for "/"
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "Success")
+}
+
+// Handler for group versions summarizer.
+func summarizeGroupVersionsHandler(path string) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var apiGroupList *unversioned.APIGroupList
+		for _, serverAddress := range groupVersionPaths[path] {
+			groupList, err := getAPIGroupList(serverAddress + path)
+			if err != nil {
+				writeErr(http.StatusInternalServerError, err, w)
+				return
+			}
+			if apiGroupList == nil {
+				apiGroupList = &unversioned.APIGroupList{}
+				*apiGroupList = *groupList
+			} else {
+				apiGroupList.Groups = append(apiGroupList.Groups, groupList.Groups...)
+			}
+		}
+		writeRawJSON(http.StatusOK, *apiGroupList, w)
+		return
+	}
+}
+
+// Handler for legacy versions summarizer.
+func summarizeLegacyVersionsHandler(path string) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if len(legacyVersionPaths[path]) > 1 {
+			err := fmt.Errorf("invalid multiple servers serving legacy group %v", legacyVersionPaths[path])
+			writeErr(http.StatusInternalServerError, err, w)
+		}
+		serverAddress := legacyVersionPaths[path][0]
+		apiVersions, err := getAPIVersions(serverAddress + path)
+		if err != nil {
+			writeErr(http.StatusInternalServerError, err, w)
+			return
+		}
+		writeRawJSON(http.StatusOK, apiVersions, w)
+		return
+	}
+}
+
+func getAPIGroupList(serverAddress string) (*unversioned.APIGroupList, error) {
+	response, err := http.Get(serverAddress)
+	if err != nil {
+		return nil, fmt.Errorf("Error in fetching %s: %v", serverAddress, err)
+	}
+	defer response.Body.Close()
+	contents, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading response from %s: %v", serverAddress, err)
+	}
+	var apiGroupList unversioned.APIGroupList
+	err = json.Unmarshal(contents, &apiGroupList)
+	if err != nil {
+		return nil, fmt.Errorf("Error in unmarshalling response from server %s: %v", serverAddress, err)
+	}
+	return &apiGroupList, nil
+}
+
+func getAPIVersions(serverAddress string) (*unversioned.APIVersions, error) {
+	response, err := http.Get(serverAddress)
+	if err != nil {
+		return nil, fmt.Errorf("Error in fetching %s: %v", serverAddress, err)
+	}
+	defer response.Body.Close()
+	contents, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading response from %s: %v", serverAddress, err)
+	}
+	var apiVersions unversioned.APIVersions
+	err = json.Unmarshal(contents, &apiVersions)
+	if err != nil {
+		return nil, fmt.Errorf("Error in unmarshalling response from server %s: %v", serverAddress, err)
+	}
+	return &apiVersions, nil
+}
+
+func writeRawJSON(statusCode int, object interface{}, w http.ResponseWriter) {
```

Comment:
I don't like that these have to be duplicated - add a TODO that common response negotiation and version handling should be split out of pkg/apiserver into a more fundamental package.